### PR TITLE
Aternity APM Prometheus Remote Storage Adapter

### PR DESCRIPTION
**Pull request by yochenAternity**
initial release of the Aternity APM Prometheus Remote Storage Adapter

### CODE REVIEW

File: prometheus/cmx/client.go [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+		level.Info(c.logger).Log("msg", "error do new metric definitions", "err", err)
+		return err
+	}
+	defer func() {
+		io.Copy(ioutil.Discard, resp.Body)
+		resp.Body.Close()
+	}()
+
+	// API returns status code 200 for successful writes.
+	// http://opentsdb.net/docs/build/html/api_http/put.html
```
Comment (spendletonAternity) [inline prometheus/cmx/client.go]: we can take out the references to opentsdb here

File: prometheus/cmx/client.go [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+	// post samples
+	if len(reqs.MetricSamples) > 0 {
+		resp2, err2 := c.putMetricSamples(reqs, u)
+
+		if err2 != nil {
+			return err2
+		}
+
+		// API returns status code 200 for successful writes.
+		// http://opentsdb.net/docs/build/html/api_http/put.html
```
Comment (spendletonAternity) [inline prometheus/cmx/client.go]: we can take out the opentsdb references

File: prometheus/main.go [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	influx "github.com/influxdata/influxdb/client/v2"
```
Comment (spendletonAternity) [inline prometheus/main.go]: do we need influxdb references here? I am not sure. Is it just an artifact from where I adapted from the influxdb adapter, or is it needed?

File: prometheus/main.go [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/common/promlog"
+	"github.com/prometheus/common/promlog/flag"
+	"gopkg.in/alecthomas/kingpin.v2"
+
+	"github.com/Aternity/custom-metrics-examples/prometheus/cmx"
+	"github.com/prometheus/prometheus/documentation/examples/remote_storage/remote_storage_adapter/graphite"
+	"github.com/prometheus/prometheus/documentation/examples/remote_storage/remote_storage_adapter/influxdb"
+	"github.com/prometheus/prometheus/documentation/examples/remote_storage/remote_storage_adapter/opentsdb"
```
Comment (spendletonAternity) [inline prometheus/main.go]: do we need opentsdb

File: prometheus/main.go [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/common/promlog"
+	"github.com/prometheus/common/promlog/flag"
+	"gopkg.in/alecthomas/kingpin.v2"
+
+	"github.com/Aternity/custom-metrics-examples/prometheus/cmx"
+	"github.com/prometheus/prometheus/documentation/examples/remote_storage/remote_storage_adapter/graphite"
```
Comment (spendletonAternity) [inline prometheus/main.go]: do we need graphite?

File: prometheus/main.go [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+	"github.com/prometheus/prometheus/documentation/examples/remote_storage/remote_storage_adapter/influxdb"
```
Comment (spendletonAternity) [inline prometheus/main.go]: do we need to reference the influxdb remote storage adapter?

File: prometheus/main.go [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+		Default("").StringVar(&cfg.graphiteAddress)
+	a.Flag("graphite-transport", "Transport protocol to use to communicate with Graphite. 'tcp', if empty.").
+		Default("tcp").StringVar(&cfg.graphiteTransport)
+	a.Flag("graphite-prefix", "The prefix to prepend to all metrics exported to Graphite. None, if empty.").
+		Default("").StringVar(&cfg.graphitePrefix)
+	a.Flag("opentsdb-url", "The URL of the remote OpenTSDB server to send samples to. None, if empty.").
+		Default("").StringVar(&cfg.opentsdbURL)
+	a.Flag("influxdb-url", "The URL of the remote InfluxDB server to send samples to. None, if empty.").
+		Default("").StringVar(&cfg.influxdbURL)
+	a.Flag("influxdb.retention-policy", "The InfluxDB retention policy to use.").
```
Comment (spendletonAternity) [inline prometheus/main.go]: do we need influxdb?

File: prometheus/main.go [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+	var writers []writer
+	var readers []reader
+	if cfg.graphiteAddress != "" {
+		c := graphite.NewClient(
+			log.With(logger, "storage", "Graphite"),
+			cfg.graphiteAddress, cfg.graphiteTransport,
+			cfg.remoteTimeout, cfg.graphitePrefix)
+		writers = append(writers, c)
+	}
+	if cfg.opentsdbURL != "" {
```
Comment (spendletonAternity) [inline prometheus/main.go]: do we need to handle these other clients? or is this only an adapter for CMX? I am not sure how the model works.

**Review by spendletonAternity — COMMENTED** [review]
do we need the references to opentsdb, influxdb and graphite in here? I am not sure of the model that the remote adapter needs. Can we just write it to support CMX, or doesn't it have to handle influxdb, graphite, etc as well?

File: prometheus/cmx/client.go (hunk 2 above)
Comment (yochenAternity) [reply prometheus/cmx/client.go]: Addressed.

File: prometheus/main.go (hunk 3 above)
Comment (yochenAternity) [reply prometheus/main.go]: Addressed.

File: prometheus/main.go (hunk 4 above)
Comment (yochenAternity) [reply prometheus/main.go]: Addressed.

File: prometheus/main.go (hunk 5 above)
Comment (yochenAternity) [reply prometheus/main.go]: Addressed.

File: prometheus/main.go (hunk 7 above)
Comment (yochenAternity) [reply prometheus/main.go]: Addressed.

File: prometheus/main.go (hunk 8 above)
Comment (yochenAternity) [reply prometheus/main.go]: Addressed.

**Comment by yochenAternity** [comment]
Removed all non Aternity APM adapters. After all, customer can always run N remote storage adapter as they desire.

The Prometheus config will look like:

```
remote_write:
    - url: ADAPTER1_URL
    - url: ADAPTER2_URL
    - ....
```

**Review by spendletonAternity — APPROVED** [review]

File: prometheus/cmx/README.md [hunk 9]
```diff
@@ -0,0 +1,7 @@
+## To run
+
+docker build --tag cmx-remote-storage-adapter:1.0 .
+
+docker run --publish 9201:9201 --name cmx-remote-storage-adapter cmx-remote-storage-adapter:1.0
+
+## TODO: make cmx/ in a seperate repo when we want to make it public on github
```
Comment (spendletonAternity) [inline prometheus/cmx/README.md]: not sure if you are still doing this?

File: prometheus/README.md [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+https://github.com/prometheus/prometheus/tree/master/documentation/examples/remote_storage/remote_storage_adapter
+
+# Remote storage adapter
+
+This is a write adapter that receives samples via Prometheus's remote write
+protocol and stores them in Aternity APM. It is meant as a
+replacement for the built-in specific remote storage implementations that have
+been removed from Prometheus.
+
+## Prereqsuisite
```
Comment (spendletonAternity) [inline prometheus/README.md]: typo Prerequisite.

File: prometheus/cmx/README.md (hunk 9 above)
Comment (yochenAternity) [reply prometheus/cmx/README.md]: addressed

File: prometheus/README.md (hunk 10 above)
Comment (yochenAternity) [reply prometheus/README.md]: addressed

File: prometheus/cmx/client.go (hunk 1 above)
Comment (yochenAternity) [reply prometheus/cmx/client.go]: addressed

File: prometheus/main.go (hunk 6 above)
Comment (yochenAternity) [reply prometheus/main.go]: addressed

**Pull request merged**